consensus/src/marshal/mod.rs

```diff
@@ -498,6 +496,166 @@ mod tests {
+            for (i, validator) in participants.iter().enumerate() {
+                let (application, actor) = setup_validator(
+                    context.with_label(&format!("validator-{i}")),
+                    &mut oracle,
+                    validator.clone(),
+                    schemes[i].clone().into(),
+                )
+                .await;
+                applications.insert(validator.clone(), application);
+                actors.push(actor);
+            }
+
+            // Add links between all peers except for the first, to guarantee
+            // the first peer does not receive any blocks during broadcast.
+            setup_network_links(&mut oracle, &participants[1..], LINK).await;
+
+            // Generate blocks, skipping the genesis block.
+            let mut blocks = Vec::<B>::new();
+            let mut parent = Sha256::hash(b"");
+            for i in 1..=NUM_BLOCKS {
+                let block = B::new::<Sha256>(parent, i, i);
+                parent = block.digest();
+                blocks.push(block);
+            }
+
+            // Broadcast and finalize blocks
+            for block in blocks.iter() {
+                // Skip genesis block
+                let height = block.height();
+                assert!(height > 0, "genesis block should not have been generated");
+
+                // Calculate the epoch and round for the block
+                let epoch = utils::epoch(BLOCKS_PER_EPOCH, height);
+                let round = Round::new(epoch, height);
+
+                // Broadcast block by one validator
+                let actor_index: usize = (height % ((NUM_VALIDATORS - 1) as u64)) as usize + 1;
+                let mut actor = actors[actor_index].clone();
+                actor.broadcast(block.clone()).await;
+                actor.verified(round, block.clone()).await;
+
+                // Wait for the block to be broadcast, but due to jitter, we may or may not receive
+                // the block before continuing.
+                context.sleep(LINK.latency).await;
+
+                // Notarize block by the validator that broadcasted it
+                let proposal = Proposal {
+                    round,
+                    parent: height.checked_sub(1).unwrap(),
+                    payload: block.digest(),
+                };
+                let notarization = make_notarization(proposal.clone(), &schemes, QUORUM);
+                actor
+                    .report(Activity::Notarization(notarization.clone()))
+                    .await;
+
+                // Finalize block by all validators except for the first.
+                let fin = make_finalization(proposal, &schemes, QUORUM);
+                for actor in actors.iter_mut().skip(1) {
+                    actor.report(Activity::Finalization(fin.clone())).await;
+                }
+            }
+
+            // Check that all applications (except for the first) received all blocks.
```

Comment:
You can now wait to create `actor` until after this point but 🤷

Reply:
Yeah, we could - done.